### ISSUE TITLE
Segfault while exporting

### ISSUE THREAD

**Issue by tomaszg7** [darktable-org/darktable#5190]
**Describe the bug**

After last rebuild of git HEAD Darktable crashes for me very often. It seems to happen after I press ^e to export image while in darkroom view. The file which I was processing doesn't get exported at all. It happens every 3-4 photos. Attaching backtrace log.

**To Reproduce**
1. Open a few images
2. Switch to darkroom mode
3. Press ^e to export

**Platform (please complete the following information):**
 - Darktable Version: 593f29a92b05d19284e517b4e7035e51b3a9b6b6 or 1deb21385f0ee97c57af37b9b4586d287c095658
 - OS: Gentoo Linux
 - OpenCL activated
 - AMD Fury + Mesa 20.1.0_rc3 + ROCm 3.3.0

**Additional context**
 - Can you reproduce with another Darktable version: Tried two recent git versions.
 - Can you reproduce with a RAW or Jpeg or both: Tried only on RAW
 - Are the steps above reproduce with a fresh edit (removing history): It happens on freshly imported images
 - Did you compile Darktable yourself? Yes, gcc-9.3.0
 - Build log: [darktable_bt_OJQ8K0.txt.gz](https://github.com/darktable-org/darktable/files/4674062/darktable_bt_OJQ8K0.txt.gz)

**Comment by johnny-bit** [comment]
can you try doing git bisect on which version causes breakages? It would help tremendously since both version you mention offer trivial changes :/

Also - any edits need to be done in order to trigger this or can it be no-edits-at-all?

**Comment by tomaszg7** [comment]
Well, that is embarrassing. After I had it crash at least four times in half an hour I submitted this bug. I've been working with DT ever since and apparently it didn't crash. I just processed 10 more photos to be sure and everything was fine. I don't think I changed anything in my system in that time. Maybe my computer is simply unstable, I just recently replaced the motherboard and maybe something is wrong. However I didn't notice any other problem today.

**Comment by johnny-bit** [comment]
Can you stress-test your system and see whether it's system or dt?

I've taken a closer look at the log you've provided and it's interesting: crash happens in `dt_history_get_items_as_string` of `history.c` which in turn gets called by thumbnail code for generating modules used tooltips. It seems like translations issue or something like that. Do you use module that isn't named/translated properly? Given how it's thumbnail tooltip code, probable way to replicate it would be to dump all modules in history and see which causes crash ;) if none then there might be different problem.

**Comment by tomaszg7** [comment]
I use DT in English, so I don't think it is related to translations. Indeed, DT didn't crash immediately after I started export, it took a short time and I might have hovered the mouse on the next photo thumbnail in that time.

Yesterday, I did try to do a naive stress test: last photos were processed when I was recompiling thunderbird in the background :) I'll try to reproduce the bug again.

**Comment by johnny-bit** [comment]
Hmmm... tha only thing that could fail there is weird. If you have sqlite3 browser check in `library.db` whether any image has any `operation`=`NULL` in history table.

I've looked at it and it seems like the issue is caused by passing `NULL` as key value to `g_str_hash` and the only cause for it would be if operation was `NULL` (or INT 0).

Since passing `NULL` as key into `g_str_hash` is not good for it, simple fix would be to disallow `NULL` op in one way or another.

**Comment by tomaszg7** [comment]
My SQL is a bit rusty but 
```
sqlite> SELECT * from history where operation=NULL;
sqlite> SELECT * from legacy_presets where operation=NULL;
```
gives no results. However the issue was in some sense self-correcting, i.e. I got a crash on one image but after restarting DT I was able to process it and the next crash occured a few images down the line. Thus it is possible that the database error corrected itself after restart.

I had a problem with operation=NULL some time ago: https://github.com/darktable-org/darktable/issues/4258

**Comment by johnny-bit** [comment]
Not to jinx it but you **might** have a faulty drive or other issues.

I've checked database.c code and it looks like `operation` column is allowed to be null, which contrasts with it's unchecked use when trying to get hash so... making sure that `operation` column is not null would be good AND also checking whether we're trying to pass NULL as key.

as for your SQL try `WHERE operation IS NULL`

**Comment by tomaszg7** [comment]
The same result with your version of SQL.

Now I recall that yesterday, when I started DT before the crashes I was prompted that the update of the db is needed. The problematic photos were added to db before that. So, it might be some db migration issue. Later I was working on newly added photos and experienced no problem.

I don't see any obvious problems with my drive, smart information looks clean (but I'm not sure if it works correctly with nvme drives).
